provides.py

```diff
@@ -15,5 +15,7 @@ def departed_broken(self):
         self.remove_state('{relation_name}.connected')
 
     def provide_data(self, tcp_port, udp_port):
-        self.set_remote(data={'tcp_port': tcp_port,
-                              'udp_port': udp_port})
+        for conv in self.conversations():
+            unit = conv.scope
+            self.set_remote(scope=unit, data={'tcp_port': tcp_port,
```

Comment:
Naming this variable "unit" is a bit misleading. Just pass the `conv.scope` directly to the set_remote() method.